extensions/assistedinject/src/com/google/inject/assistedinject/FactoryProvider2.java

```diff
@@ -894,54 +895,83 @@ protected Object initialValue() {
     }
   }
 
-  // Note: this isn't a public API, but we need to use it in order to call default methods on (or
-  // with) non-public types.  If it doesn't exist, the code falls back to a less precise check.
-  private static final Constructor<MethodHandles.Lookup> methodHandlesLookupCxtor =
-      findMethodHandlesLookupCxtor();
-
-  private static Constructor<MethodHandles.Lookup> findMethodHandlesLookupCxtor() {
-    // Different JDK implementations have different constructors so look for a constructor that
-    // takes a class and an int first (openjdk-8, openjdk-11) and fallback to a constructor that
-    // takes two classes and an int (openjdk-15).
-    // TODO(b/171738889): Figure out a better way to handle this.
-    Constructor<MethodHandles.Lookup> cxtor = findMethodHandlesLookupCxtor(Class.class, int.class);
-    if (cxtor == null) {
-      cxtor = findMethodHandlesLookupCxtor(Class.class, Class.class, int.class);
-    }
-    if (cxtor != null) {
-      cxtor.setAccessible(true);
-    }
-    return cxtor;
-  }
-
-  private static Constructor<MethodHandles.Lookup> findMethodHandlesLookupCxtor(
-      Class<?>... parameterTypes) {
+  private static MethodHandle superMethodHandle(Method method, Object proxy) {
     try {
-      return MethodHandles.Lookup.class.getDeclaredConstructor(parameterTypes);
-    } catch (NoSuchMethodException e) {
-      // Ignored if the constructor doesn't exist.
-      return null;
+      return SUPER_METHOD_LOOKUP.superMethodHandle(method).bindTo(proxy);
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to access method: " + method, e);
     }
   }
 
-  private static MethodHandle createMethodHandle(Method method, Object proxy) {
-    if (methodHandlesLookupCxtor == null) {
-      return null;
-    }
-    Class<?> declaringClass = method.getDeclaringClass();
-    int allModes =
+  // begin by trying unreflectSpecial to find super method handles; this should work on Java14+
+  private static SuperMethodLookup SUPER_METHOD_LOOKUP = SuperMethodLookup.UNREFLECT_SPECIAL;
+
+  private static enum SuperMethodLookup {
+    UNREFLECT_SPECIAL {
+      @Override
+      MethodHandle superMethodHandle(Method method) throws Exception {
+        try {
+          return MethodHandles.lookup().unreflectSpecial(method, method.getDeclaringClass());
+        } catch (Exception e) {
+          // fall back to findSpecial which should work on Java9+; use that for future lookups
+          SUPER_METHOD_LOOKUP = FIND_SPECIAL;
+          return FIND_SPECIAL.superMethodHandle(method);
+        }
+      }
+    },
+    FIND_SPECIAL {
+      @Override
+      MethodHandle superMethodHandle(Method method) throws Exception {
+        try {
+          Class<?> declaringClass = method.getDeclaringClass();
+          // use findSpecial to workaround https://bugs.openjdk.java.net/browse/JDK-8209005
+          return MethodHandles.lookup().findSpecial(declaringClass, method.getName(),
```

Comment:
OOC, in what ways does findSpecial differ from unreflectSpecial?   are the two functionally equivalent (ignoring the JDK bugs), except one stards with reflect data (Method) and the other with a signature (MethodType)?

if so, is it worthwhile simplifying this to _just_ use findSpecial?

also: does this change have any impact on the calculus of who has access?  if the types (args or return) are private, package-private or protected, will unreflectSpecial|findSpecial fail whereas the reflected approach would have succeeded, because we explicitly set the modes to all access levels?

Reply:
`unreflectSpecial` is intended to be equivalent to `findSpecial` but slightly faster because it avoids an extra resolution step. (The JVM can re-use detail in the reflected method so it doesn't need to match the given name+type signature against potential candidates.) It also skips security checks which have already been performed during reflection.

Just using `findSpecial` would save one call to `unreflectSpecial` at the cost of always re-resolving default methods on Java14+... note it would only save one call because if `unreflectSpecial` fails then we remember that and avoid trying it again.

I'm tempted to leave it in because I don't think removing it would simplify the code that much - we'd still want to remember falling back to the private lookup if/when that happens because otherwise we'd penalise Java8 users by always attempting a different approach even when we knew it would never work. (That's if we still follow the same general pattern - did you have a different suggestion?)

Regarding access, since this PR always falls back to the `privateLookup` approach if all of the other approaches fail then this shouldn't change existing behaviour. What it does do is avoid the `privateLookup` if any of the other approaches succeed first (which is more likely on recent JDKs.)

Reply:
Leaving it in if it makes a difference (even it's just a single resolution step) makes sense to me.

Insofar as falling back to privateLookup because of access issues... that makes less sense to me, particularly because the fallback mechanism is static.  So if using unreflectSpecial|findSpecial fails for non-public members, then if you have just a single non-public usage in your code, you can expect all your code to ultimately be using the reflection-based approach and continuing to receive the warnings about that.

That's not to say that I think we should use try the tiered approach every time... but it does make me wonder about the utility of it

Reply:
Fundamentally we either take a versioned approach, using the JDK version to decide what to do - or we take a behaviour-based approach, starting with the available public API and falling back to internal API when that fails. Because it would be expensive to continually fall-through the different approaches on Java 8, this PR caches (statically) which approach worked for all the scenarios so far.

IMHO this is reasonable because if we've fallen through to use the private lookup then it's likely this will happen again. At that point we've already done the hard bit (looking up the constructor and making it accessible) so re-using the now accessible constructor is simple and doesn't trigger further warnings. In other words since we've got access to it we might as well skip straight to it.

Also in this case we know we're dealing with public methods, because default methods are always public. This coupled with the `unreflectSpecial` fixes in Java 14+ means that approach works even when the interface itself is private. If the JVM is older then we fall back to `findSpecial` which works for Java9+ even when the interface itself is private. Finally for Java8 we fall back to the internal API.

Regarding the different `Lookup` constructor, that appeared in Java14 - so far I haven't found a situation where `unreflectSpecial` fails on Java14 even for private interfaces, but I haven't tested all JPMS scenarios. But we can easily add support for the Java14+ Lookup constructor just in case. The alternative would be to grab the trusted `IMPL_LOOKUP` field, but that feels more risky.